### PR TITLE
Display weather forecast with icons

### PR DESCRIPTION
**Pull request by gelbal**
To make it backwards compatible, PR adds extra environment variables. So the user needs to customize the configuration for Weather display to take precedence over Quote display.
Use free APIs which don't require authentication to retrieve location and weather data.
Refactor the original code logic a bit.

Change-type: major
Signed-off-by: Fırat Gelbal firat@balena.io

### CODE REVIEW

File: start.sh [hunk 1]
```diff
@@ -11,8 +11,17 @@ python /usr/app/update-display.py
 # Save out the current env to a file so cron job can use it
 export -p > /usr/app/env.sh
 
+# Set default values if these env vars are not set
```
Comment (gelbal) [inline start.sh:14]: Added more complex logic here to support display alternating and more flexible refresh frequency.

File: start.sh [hunk 2]
```diff
@@ -11,8 +11,17 @@ python /usr/app/update-display.py
...
+  Alternate="0"
+  [[ -z "${UPDATE_HOUR}" ]] && UpdateHour='9' || UpdateHour="${UPDATE_HOUR}"
+else
+  Alternate="*/${ALTERNATE_FREQUENCY}"
+  UpdateHour='*'
+fi
+
 # Add the job to the crontab using update_hour var, defaulting to 9 AM
-(echo "0 ${UPDATE_HOUR:-9} * * * /usr/app/run-update.sh") | crontab -
+(echo "${Alternate} ${UpdateHour} * * * /usr/app/run-update.sh") | crontab -
```
Comment (gelbal) [inline start.sh:24]: For some reason `cron` job does not trigger on my device, although I see that it's set properly. I also tested a custom cron schedule to update every 60 minutes (`*/60 * * * * /usr/app/run-update.sh`) on the master code and that also did not run. Although the original code runs fine on daily trigger.

File: update-display.py [hunk 3]
```diff
@@ -1,45 +1,222 @@
+import json
```
Comment (gelbal) [inline update-display.py:1]: Re-ordered the imports because of order obsession : )

File: update-display.py [hunk 4]
```diff
@@ -1,45 +1,222 @@
...
-from font_caladea import Caladea
 from font_roboto import Roboto
-from font_amatic_sc import AmaticSC
+from font_source_sans_pro import SourceSansPro
+from font_source_serif_pro import SourceSerifPro
+from PIL import Image, ImageFont, ImageDraw
+import arrow
+import geocoder
+
+icon_map = {
```
Comment (gelbal) [inline update-display.py:24]: Respects the weather icon numbers from the upstream source.

File: update-display.py [hunk 5]
```diff
@@ -1,45 +1,222 @@
...
+    "sleetandthunder": 23,
+    "sleetshowers": 7,
+    "sleetshowersandthunder": 20,
+    "snow": 13,
+    "snowandthunder": 14,
+    "snowshowers": 8,
+    "snowshowersandthunder": 21,
+}
+
+def create_mask(source):
```
Comment (gelbal) [inline update-display.py:68]: Right now the weather icons and masks are generated for Black / White displays. Then we could easily extend this to show other colors the screen supports.

**Review by gelbal — COMMENTED** [review]
Hey @chrisys @catt-tung I appreciate your review and your test. This turned out to be a larger change than I expected but it's been fun : )

The only problem is the `cron` job does not trigger. I couldn't figure out the problem so it'd be great if you could look into it. The cron job of the latest code in production runs fine daily on my device. But when I set it to run every 60 minutes, neither does that trigger. So I suspect that it's a configuration problem somewhere (or I mistype the cron job change).

File: start.sh (hunk 2 above)
Comment (catt-tung) [reply start.sh:24]: Hm @chrisys are you able to test if this works on yours? I wonder if it has anything to do with the display being different

**Comment by gelbal** [comment]
Hey @chrisys @catt-tung could you please test this on your devices?

After rebasing on master and applying the change from https://github.com/balenalabs/inkyshot/pull/25, the cron job started working fine. I also fixed a bug in the code. The code with weather changes seems like working fine now :tada: 

I'd like to get your review and merge the changes.

**Pull request merged**